internal/dataplane/sendconfig/backoff_strategy.go

```diff
@@ -0,0 +1,81 @@
+package sendconfig
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/adminapi"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/metrics"
+)
+
+type ErrUpdateSkippedDueToBackoffStrategy struct {
+	explanation string
+}
+
+func NewErrUpdateSkippedDueToBackoffStrategy(explanation string) ErrUpdateSkippedDueToBackoffStrategy {
+	return ErrUpdateSkippedDueToBackoffStrategy{explanation: explanation}
+}
+
+func (e ErrUpdateSkippedDueToBackoffStrategy) Error() string {
+	return fmt.Sprintf("update skipped due to a backoff strategy not being satisfied: %s", e.explanation)
+}
+
+func (e ErrUpdateSkippedDueToBackoffStrategy) Is(err error) bool {
+	return errors.Is(err, ErrUpdateSkippedDueToBackoffStrategy{})
+}
+
+// UpdateStrategyWithBackoff decorates any UpdateStrategy to respect a passed adminapi.UpdateBackoffStrategy.
+type UpdateStrategyWithBackoff struct {
+	decorated       UpdateStrategy
+	backoffStrategy adminapi.UpdateBackoffStrategy
+	log             logrus.FieldLogger
+}
+
+func NewUpdateStrategyWithBackoff(
+	decorated UpdateStrategy,
+	backoffStrategy adminapi.UpdateBackoffStrategy,
+	log logrus.FieldLogger,
+) UpdateStrategyWithBackoff {
+	return UpdateStrategyWithBackoff{
+		decorated:       decorated,
+		backoffStrategy: backoffStrategy,
+		log:             log,
+	}
+}
+
+// Update will ensure that the decorated UpdateStrategy.Update is called only when an underlying
+// UpdateBackoffStrategy.CanUpdate is satisfied.
+// In case it's not, it will return a predefined ErrUpdateSkippedDueToBackoffStrategy.
+// In case it is, apart from calling UpdateStrategy.Update, it will also register a success or a failure of an update
+// attempt so that the UpdateBackoffStrategy can keep track of it.
+func (s UpdateStrategyWithBackoff) Update(ctx context.Context, targetContent ContentWithHash) (
+	err error,
+	resourceErrors []ResourceError,
+	resourceErrorsParseErr error,
+) {
+	if canUpdate, whyNot := s.backoffStrategy.CanUpdate(targetContent.Hash); !canUpdate {
+		s.log.Debug("Skipping update due to a backoff strategy")
```

Comment:
Perhaps log the `whyNot`?

Reply:
Hm, actually this log is redundant as the update result is logged up in the `KongClient`, where we have the `ErrUpdateSkippedDueToBackoffStrategy` propagated.

I also changed the log levels from error to warning as Konnect config push failure is not strictly an error as it doesn't affect core functionality of the controller. PTAL if that makes sense to you.